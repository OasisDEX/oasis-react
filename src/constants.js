/**
 *   NETWORKS
 */
export const MAIN_NET_ID = "1";
export const KOVAN_NET_ID = "42";
export const KOVAN = "kovan";
export const MAIN = "main";

/**
 * NETWORK STATUS
 */
export const ONLINE = "NETWORK/ONLINE";
export const CONNECTING = "NETWORK/CONNECTING";
export const OUT_OF_SYNC = "NETWORK/OUT_OF_SYNC";
export const CLOSED = "NETWORK/CLOSED";

/**
 * BIG NUMBER
 */
export const BN_DECIMAL_PRECISION = 18;

/**
 * PRECISION
 */

export const ETH_UNIT_WEI = "wei"; // wei 	1 wei 	1
export const ETH_UNIT_KWEI = "kwei"; // Kwei (babbage) 	1e3 wei 	1,000
export const ETH_UNIT_MWEI = "mwei"; // Mwei (lovelace) 	1e6 wei 	1,000,000
export const ETH_UNIT_GWEI = "gwei"; // Gwei (shannon) 	1e9 wei 	1,000,000,000
export const ETH_UNIT_MICROETHER = "microether"; // microether (szabo) 	1e12 wei 	1,000,000,000,000
export const ETH_UNIT_MILIETHER = "milliether"; // milliether (finney) 	1e15 wei 	1,000,000,000,000,000
export const ETH_UNIT_ETHER = "ether"; // ether 	1e18 wei 	1,000,000,000,000,000,000

export const HAS_ACCOUNTS = true;
export const NO_ACCOUNTS = false;

export const TOKEN_ETHER = "ETH";
export const TOKEN_WRAPPED_ETH = "W-ETH";
export const TOKEN_WRAPPED_GNT = "W-GNT";
export const TOKEN_DAI = "DAI";
export const TOKEN_MAKER = "MKR";
export const TOKEN_DIGIX = "DGD";
export const TOKEN_GOLEM = "GNT";
export const TOKEN_RHOC = "RHOC";

export const QUOTE_TOKENS = [TOKEN_WRAPPED_ETH, TOKEN_DAI];

export const BASE_TOKENS = [
  TOKEN_DAI,
  TOKEN_WRAPPED_GNT,
  TOKEN_DIGIX,
  TOKEN_RHOC,
];

export const MAKE_BUY_OFFER = "OFFER_MAKES/MAKE_BUY_OFFER";
export const MAKE_SELL_OFFER = "OFFER_MAKES/MAKE_SELL_OFFER";
export const MAKE_SELL_OFFER_FORM_NAME = "makeSellOffer";
export const MAKE_BUY_OFFER_FORM_NAME = "makeBuyOffer";

export const TOKEN_ALLOWANCE_TRUST_STATUS_DISABLED_MIN_MAX = 0;
export const TOKEN_ALLOWANCE_TRUST_STATUS_ENABLED = true;
export const TOKEN_ALLOWANCE_TRUST_STATUS_DISABLED = false;
export const TOKEN_ALLOWANCE_TRUST_STATUS_NOT_SET = undefined;
export const TOKEN_ALLOWANCE_TRUST_STATUS_LOADING = null;
export const TOKEN_ALLOWANCE_TRUST_STATUS_ENABLED_MIN =
  "0xffffffffffffffffffffffffffffffff";
export const TOKEN_ALLOWANCE_TRUST_STATUS_ENABLED_MAX =
  "0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff";

export const TOKEN_ALLOWANCE_TRUST_SUBJECT_TYPE_MARKET =
  "BALANCES/TOKEN_ALLOWANCE_TRUST_SUBJECT_TYPE_MARKET";
export const TOKEN_ALLOWANCE_TRUST_SUBJECT_TYPE_ADDRESS =
  "BALANCES/TOKEN_ALLOWANCE_TRUST_SUBJECT_TYPE_ADDRESS";

export const SYNC_STATUS_PRISTINE = "SYNC_STATUS_PRISTINE";
export const SYNC_STATUS_PENDING = "SYNC_STATUS_PENDING";
export const SYNC_STATUS_COMPLETED = "SYNC_STATUS_COMPLETED";
export const SYNC_STATUS_ERROR = "SYNC_STATUS_ERROR";

export const SETMAXBTN_HIDE_DELAY_MS = 100;

/**
 *  GAS and GAS_PRICE default values
 */
export const DEFAULT_GAS_LIMIT = "1000000"; // 1M GAS
export const DEFAULT_GAS_PRICE = "6000000000"; // 6 GWEI

/**
 * Subscriptions
 */

export const SUBSCRIPTIONS_GROUP_GLOBAL_INITIAL = 'SUBSCRIPTIONS/GROUP_GLOBAL/INITIAL';
export const SUBSCRIPTIONS_GROUP_ACCOUNT_SPECIFIC_INITIAL = 'SUBSCRIPTIONS/GROUP_ACCOUNTS_SPECIFIC/INITIAL';

export const SUBSCRIPTIONS_LATEST_BLOCK = "SUBSCRIPTIONS/LATEST_BLOCK";
export const SUBSCRIPTIONS_TOKEN_BALANCE_EVENTS = "SUBSCRIPTIONS/TOKEN_BALANCES";
export const SUBSCRIPTIONS_ORDERS_EVENTS = "SUBSCRIPTIONS/ORDERS_EVENTS";
export const SUBSCRIPTIONS_ETHER_BALANCE_CHANGE_EVENTS = "SUBSCRIPTIONS/ETHER_BALANCE_CHANGE_EVENTS";
export const SUBSCRIPTIONS_LOG_TAKE_EVENTS = "SUBSCRIPTIONS/LOG_TAKE_EVENTS";
export const SUBSCRIPTIONS_USER_LOG_TAKE_EVENTS = "SUBSCRIPTIONS/USER_LOG_TAKE_EVENTS";
export const SUBSCRIPTIONS_TOKEN_TRANSFER_EVENTS = "SUBSCRIPTIONS/TOKEN_TRANSFER_EVENTS";
export const SUBSCRIPTIONS_TOKEN_TRANSFER_HISTORY_EVENTS = "SUBSCRIPTIONS/TOKEN_TRANSFER_HISTORY_EVENTS";
export const SUBSCRIPTIONS_TOKEN_WRAP_UNWRAP_HISTORY_EVENTS = "SUBSCRIPTIONS/TOKEN_TRANSFER_HISTORY_EVENTS";


export const subscriptionGroupToKeyMap = {
  [SUBSCRIPTIONS_GROUP_GLOBAL_INITIAL]: 'globalInitial',
  [SUBSCRIPTIONS_GROUP_ACCOUNT_SPECIFIC_INITIAL]: 'accountSpecificInitial'
};

export const subscriptionTypeToKeyMap = {
  [SUBSCRIPTIONS_LATEST_BLOCK]: 'latestBlock',
  // [SUBSCRIPTIONS_TOKEN_BALANCE_EVENTS]: 'tokenBalances',
  [SUBSCRIPTIONS_ORDERS_EVENTS]: 'ordersEvents',
  [SUBSCRIPTIONS_LOG_TAKE_EVENTS]: 'latestBlock',
  [SUBSCRIPTIONS_TOKEN_WRAP_UNWRAP_HISTORY_EVENTS]: 'wrapUnwrapHistory',
  [SUBSCRIPTIONS_ETHER_BALANCE_CHANGE_EVENTS]: 'etherBalance',
  [SUBSCRIPTIONS_TOKEN_TRANSFER_EVENTS]: 'tokenTransfers'
};

/**
 * Validation errors
 */

export const VALIDATION_VALUE_IS_REQUIRED = 'VALIDATION/VALUE_IS_REQUIRED';

/**
 * Global precision - 10**18
 * @type {number}
 */
export const APP_BASE_PRECISION = 18;

export const USER_TO_LOG_TAKE_OFFER_RELATION_TAKEN_BY_USER = 'USER_TO_LOG_TAKE_RELATION/TAKEN_BY_USER';
export const USER_TO_LOG_TAKE_OFFER_RELATION_USER_MADE = 'USER_TO_LOG_TAKE_RELATION/USER_MADE';
export const USER_TO_LOG_TAKE_OFFER_RELATION_NONE = 'USER_TO_LOG_TAKE_RELATION/NONE';

export const PROVIDER_TYPE_METAMASK = 'PROVIDER_TYPE/METAMASK';
export const PROVIDER_TYPE_MIST = 'PROVIDER_TYPE/MIST';
