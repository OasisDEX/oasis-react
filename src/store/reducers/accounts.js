import { createAction, handleActions } from 'redux-actions';
import Immutable from 'immutable';

import web3 from '../../bootstrap/web3';
import { Session } from '../../utils/session';
import platformReducer from './platform';
import sessionReducer from './session';


const initialState = Immutable.fromJS({
  defaultAccount: null,
  accounts: []
});

const INIT = 'ACCOUNTS/INIT';
const CHECK_ACCOUNTS = 'NETWORK/CHECK_ACCOUNTS';

const Init = createAction(
  INIT,
  () => null,
);



const checkAccounts = createAction(
  CHECK_ACCOUNTS,
  () => window.web3p.eth.getAccounts(),
);


const checkAccountsEpic = () => async (dispatch, getState) => {
  const userAccounts = (await dispatch(checkAccounts())).value;
  const isMetamaskLocked = getState().getIn(['platform', 'metamaskLocked']);
  const hasUserAccounts = userAccounts.length;
  if(!hasUserAccounts) {
    if(!isMetamaskLocked) {
      dispatch(platformReducer.actions.metamaskLocked());
    }
    window.web3.eth.defaultAccount = undefined;
  }
  else if (!userAccounts.find( acc => acc === window.web3.eth.defaultAccount)) {
    if ((userAccounts.find( acc => acc === localStorage.getItem('address')) )) {
      window.web3.eth.defaultAccount = localStorage.getItem('address');
    } else if (userAccounts.find( acc => acc === Session.get('address'))) {
      window.web3.eth.defaultAccount = Session.get('address');
    } else if (userAccounts.length > 0) {
      window.web3.eth.defaultAccount = userAccounts[0];
    } else {
      throw new Error('No default account');
    }
  }

  if(isMetamaskLocked) {
    dispatch(platformReducer.actions.metamaskUnlocked())
  }
  localStorage.setItem('address', web3.eth.defaultAccount);
  dispatch(sessionReducer.actions.SetValue('address', window.web3.eth.defaultAccount));
  dispatch(sessionReducer.actions.SetValue('accounts', userAccounts))
};


const actions = {
  Init,
  checkAccountsEpic
};

const reducer = handleActions({}, initialState);

export default {
  actions,
  reducer,
};
