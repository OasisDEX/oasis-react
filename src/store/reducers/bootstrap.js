import { createAction, handleActions } from 'redux-actions';
import Immutable from 'immutable';

const initialState = Immutable.fromJS({});

const NETWORK_INITIALIZED = 'BOOTSTRAP/NETWORK_INITIALIZED';



const networkInitialized = createAction(
  NETWORK_INITIALIZED
);

const actions = {
  networkInitialized
};

const reducer = handleActions({}, initialState);

export default {
  actions,
  reducer,
};
