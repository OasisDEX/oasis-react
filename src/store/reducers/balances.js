import { createAction, handleActions } from 'redux-actions';
import Immutable from 'immutable';
import {createPromiseActions} from '../../utils/createPromiseActions';
import {fulfilled} from '../../utils/store';


const initialState = Immutable.fromJS({
  accounts: [],
  defaultAccount: {
    address: null,
    ethBalance: null,
    tokenBalances: {},
    tokenAllowances: {}
  },

});

const INIT = 'BALANCES/INIT';
const GET_DEFAULT_ACCOUNT_ETH_BALANCE = 'BALANACES/GET_DEFAULT_ACCOUNT_ETH_BALANCE';
const GET_ALL_TRADED_TOKENS_BALANCES = 'BALANACES/GET_ALL_TRADED_TOKENS_BALANCES';
const GET_ALL_TRADED_TOKENS_ALLOWANCES = 'BALANACES/GET_ALL_TRADED_TOKENS_ALLOWANCES';

const Init = createAction(
  INIT,
  () => null,
);


const getDefaultAccountEthBalance = createAction(
  GET_DEFAULT_ACCOUNT_ETH_BALANCE,
  async () => window.web3p.eth.getBalance(window.web3.eth.defaultAccount)
);



const getAllTradedTokensBalances = createAction(
  GET_ALL_TRADED_TOKENS_BALANCES,
  async (tokensContractsLists) => {

    const tokensBalancesPromises = [];

    Object.entries(tokensContractsLists).forEach(
      async ([tokenName, tokenContract]) => {
          if(tokenContract.balanceOf) {
          tokensBalancesPromises.push(
            tokenContract.balanceOf(window.web3.eth.defaultAccount)
          );
        }
      }
    );

    return Promise.all(tokensBalancesPromises).then(
      tokenBalances => {

        const balancesByToken = {};
        Object.keys(tokensContractsLists).forEach(
          (tokenName, i) => balancesByToken[tokenName] = tokenBalances[i]
        );

        return balancesByToken;
      }
    )
  }
);

const getAllTradedTokensAllowances = createAction(
  GET_ALL_TRADED_TOKENS_ALLOWANCES,
  async (tokensContractsLists, spenderAddress) => {

    const tokensAllowancesPromises = [];

    Object.entries(tokensContractsLists).forEach(
      async ([tokenName, tokenContract]) => {
        if(tokenContract.allowance) {
          tokensAllowancesPromises.push(
            tokenContract.allowance(window.web3.eth.defaultAccount, spenderAddress)
          );
        }
      }
    );

    return Promise.all(tokensAllowancesPromises).then(
      tokenAllowances => {
        const allowancesByToken = {};
        Object.keys(tokensContractsLists).forEach(
          (tokenName, i) => allowancesByToken[tokenName] = tokenAllowances[i]
        );

        return allowancesByToken;
      }
    )
  }
);



const tokenTransferEvent = createAction(
  'BALANCES/EVENT___TOKEN_TRANSFER',
  (tokenName, userAddress, event) => ({ tokenName, event })
)

const tokenTransferFromEvent = createAction(
  'BALANCES/EVENT___TOKEN_TRANSFER_FROM',
  (tokenName, userAddress, event) => ({ tokenName, userAddress, event })
)

const tokenTransferToEvent = createAction(
  'BALANCES/EVENT___TOKEN_TRANSFER_TO',
  (tokenName, userAddress, event) => ({ tokenName, userAddress, event })
);

const subscribeTokenTransfersEvents = createPromiseActions(
  'BALANCES/SUBSCRIBE_TOKEN_TRANSFER_EVENT'
);

const subscribeTokenTransfersEventsEpic = (tokensContractsList, address, config = {}) => async (dispatch) => {
  dispatch(subscribeTokenTransfersEvents.pending());
  Object.entries(tokensContractsList).forEach(
    ([tokenName, tokenContract]) => {
      /**
       * Listen to all erc20 transfer events from now.
       */
      tokenContract.Transfer(address, config)
        .then(
          transferEvent => {
            dispatch(tokenTransferEvent(tokenName, transferEvent, address));
          }
        );
    }
  );
  dispatch(subscribeTokenTransfersEvents.fulfilled());
};

const actions = {
  Init,
  getDefaultAccountEthBalance,
  getAllTradedTokensBalances,
  getAllTradedTokensAllowances,
  subscribeTokenTransfersEventsEpic
};

const reducer = handleActions({

  [fulfilled(getAllTradedTokensBalances)]: (state, action ) =>
    state.updateIn(
      ['defaultAccount', 'tokenBalances'],
      (balances) => {
        const tokenBalances = action.payload;
        Object.entries(tokenBalances).forEach(
          ([tokenName, tokenBalance]) => {
            balances = balances.set(tokenName, tokenBalance ? tokenBalance.toFormat(10) : null)
          }
        );
        return balances
      }),

  [fulfilled(getAllTradedTokensAllowances)]: (state, action ) =>
    state.updateIn(
      ['defaultAccount', 'tokenAllowances'],
      (allowances) => {
        const tokenAllowances = action.payload;
        Object.entries(tokenAllowances).forEach(
          ([tokenName, tokenAllowance]) => {
            allowances = allowances.set(tokenName, tokenAllowance.toFormat(10))
          }
        );
        return allowances
      }),
  [tokenTransferEvent]: (state, { payload: { tokenName, userAddress, event } }) => {

    console.log({tokenName, userAddress, event})
   return state.updateIn(['tokenBalances', tokenName], (balance) => balance)
  }


}, initialState);

export default {
  actions,
  reducer,
};
