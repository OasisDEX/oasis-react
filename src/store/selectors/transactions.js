import { createSelector } from "reselect";
import reselect from "../../utils/reselect";
import web3 from "../../bootstrap/web3";
import {
  getTransactionGroup,
  TX__GROUP__LIMITS,
  TX__GROUP__OFFERS,
  TX__GROUP__TOKENS,
  TX__GROUP__TRANSFERS
} from "../reducers/transactions";

import balances from "./balances";
import { DEFAULT_GAS_LIMIT } from '../../constants';

const transactions = s => s.get("transactions");

const transactionsList = createSelector(transactions, s => s.get("txList"));

const offersTransactions = createSelector(transactionsList, s =>
  s.filter(
    offer => getTransactionGroup(offer.get("txType")) === TX__GROUP__OFFERS
  )
);

const tokensTransactions = createSelector(transactionsList, s =>
  s.filter(
    offer => getTransactionGroup(offer.get("txType")) === TX__GROUP__TOKENS
  )
);
const limitsTransactions = createSelector(transactionsList, s =>
  s.filter(
    offer => getTransactionGroup(offer.get("txType")) === TX__GROUP__LIMITS
  )
);

const transferTransactions = createSelector(transactionsList, s =>
  s.filter(
    offer => getTransactionGroup(offer.get("txType")) === TX__GROUP__TRANSFERS
  )
);

const getOfferTransaction = createSelector(
  offersTransactions,
  reselect.getProps,
  (offersTransactions, { offerId }) =>
    offersTransactions.find(
      txEl => parseInt(txEl.get("txSubjectId")) === parseInt(offerId)
    )
);

const getTokenTransaction = createSelector(
  tokensTransactions,
  reselect.getProps,
  (tokensTransactionsList, { tokenId }) =>
    tokensTransactionsList.find(
      txEl => parseInt(txEl.get("txSubjectId")) === parseInt(tokenId)
    )
);

const getLimitTransaction = createSelector(
  limitsTransactions,
  reselect.getProps,
  (limitsTransactionsList, { txTimestamp }) =>
    limitsTransactionsList.find(
      txEl => txEl.getIn(["txStats", "txStartTimestamp"]) === txTimestamp
    )
);

const getTransferTransaction = createSelector(
  transferTransactions,
  reselect.getProps,
  (transfersTransactionsList, txSubjectId) =>
    transfersTransactionsList.find(
      txEl => txEl.get("txSubjectId") === txSubjectId
    )
);

const defaultGasLimit = createSelector(transactions, s =>
  s.get("defaultGasLimit")
);

const activeGasLimit = createSelector(transactions, s =>
  s.get("activeGasLimit")
);

const currentGasPriceWei = createSelector(transactions, s =>
  s.get("currentGasPriceInWei")
);

const canSendTransaction = createSelector(
  balances.ethBalance,
  reselect.getProps,
  (ethBalance, gasCost) =>
    web3.toBigNumber(ethBalance).gte(gasCost || DEFAULT_GAS_LIMIT)
);

const getTransactionByTxHash = createSelector(
  transactionsList,
  reselect.getProps,
  (txList, { txHash }) => txList.find(tx => tx.get("txHash") === txHash)
);

const getTransactionByTimestampAndType = createSelector(
  transactionsList,
  reselect.getProps,
  (txList, { txType, txTimestamp } = {}) =>
    txList.find(
      tx =>
        tx.get("txType") === txType &&
        tx.getIn(["txStats", "txStartTimestamp"]) === txTimestamp
    )
);

const currentTxNonce = createSelector(transactions, s => s.get("txNonce"));

const transactionCheckInterval = createSelector(transactions, s =>
  s.get("transactionCheckIntervalMs")
);

export default {
  state: transactions,
  activeGasLimit,
  defaultGasLimit,
  offersTransactions,
  tokensTransactions,
  limitsTransactions,
  getOfferTransaction,
  getTokenTransaction,
  getLimitTransaction,
  canSendTransaction,
  currentGasPriceWei,
  getTransactionByTxHash,
  currentTxNonce,
  getTransferTransaction,
  getTransactionByTimestampAndType,
  transactionCheckInterval
};
