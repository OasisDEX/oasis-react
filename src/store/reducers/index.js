import { combineReducers } from 'redux-immutable';
import { routerReducer } from 'react-router-redux'

import platform from './platform';
import network from './network';
import markets from './markets';
import balances from './balances';
import orders from './orders';
import history from './transferHistory';
import limits from './limits';
import offers from './offers';
import tokenEvents from './tokenEvents';
import tokens from './tokens';
import transactions from './transactions';
import weth from './weth';
import wgnt from './wgnt';
import session from './session';
import contracts from './contracts';
import accounts from './accounts';
import trades from './trades';
import widgets from './widgets';
import offerTakes from './offerTakes';
import { reducer as formReducer } from 'redux-form/immutable'
import transactionWatchers from './transactionWatchers';
import timers from './timers';
import offerMakes from './offerMakes';
import transferHistory from './transferHistory';
import transfers from './transfers';
import tokenSelectors from './tokenSelectors';
import wrapUnwrap from './wrapUnwrap';


export default combineReducers(
  {

    router: routerReducer,
    platform: platform.reducer,
    network: network.reducer,
    markets: markets.reducer,
    balances: balances.reducer,
    orders: orders.reducer,
    history: history.reducer,
    limits: limits.reducer,
    offers: offers.reducer,
    tokenEvents: tokenEvents.reducer,
    tokens: tokens.reducer,
    transactions: transactions.reducer,
    weth: weth.reducer,
    wgnt: wgnt.reducer,
    session: session.reducer,
    contracts: contracts.reducer,
    accounts: accounts.reducer,
    trades: trades.reducer,
    widgets: widgets.reducer,
    offerTakes: offerTakes.reducer,
    offerMakes: offerMakes.reducer,
    form: formReducer,
    transactionWatchers: transactionWatchers.reducer,
    timers: timers.reducer,
    transferHistory: transferHistory.reducer,
    transfers: transfers.reducer,
    tokenSelectors: tokenSelectors.reducer,
    wrapUnwrap: wrapUnwrap.reducer
  });
