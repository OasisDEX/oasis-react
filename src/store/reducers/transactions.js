import { createAction, handleActions } from 'redux-actions';
import { fromJS } from 'immutable';
import { web3p } from '../../bootstrap/web3';
import { createPromiseActions } from '../../utils/createPromiseActions';
import network from '../selectors/network';
import { fulfilled } from '../../utils/store';
import web3 from '../../bootstrap/web3';
import accounts from '../selectors/accounts';
import { getTimestamp } from '../../utils/time';
import transactions from '../selectors/transactions';
import { DEFAULT_GAS_LIMIT, DEFAULT_GAS_PRICE } from '../../constants';

const TRANSACTION_CHECK_INTERVAL_MS = 1000;

const initialState = fromJS({
  txList: [],
  defaultGasLimit: DEFAULT_GAS_LIMIT,
  activeGasLimit: DEFAULT_GAS_LIMIT,
  defaultGasPrice: DEFAULT_GAS_PRICE,
  activeGasPrice: DEFAULT_GAS_PRICE,
  currentGasPriceInWei: null,
  txNonce: null,
  transactionCheckIntervalMs: TRANSACTION_CHECK_INTERVAL_MS
});

export const TX__GROUP__OFFERS = 'TRANSACTIONS/GROUP__OFFERS/';
export const TX__GROUP__TOKENS = 'TX/GROUP__TOKENS/';
export const TX__GROUP__LIMITS = 'TX/GROUP__LIMITS/';
export const TX__GROUP__TRANSFERS = 'TX/GROUP__TRANSFERS/';
export const TX__GROUP__WRAP_UNWRAP = 'TX/GROUP__WRAP_UNWRAP/';


export const TX_OFFER_MAKE = TX__GROUP__OFFERS + 'OFFER_MAKE';
export const TX_OFFER_TAKE = TX__GROUP__OFFERS + 'OFFER_TAKE';

export const TX_OFFER_CANCEL = TX__GROUP__OFFERS + 'OFFER_CANCEL';

export const TX_OFFER_PARTIALLY_FULFILLED = TX__GROUP__OFFERS + 'OFFER_FULFILLED_PARTIALLY';
export const TX_OFFER_FULFILLED_COMPLETELY = TX__GROUP__OFFERS + 'OFFER_FULFILLED_COMPLETELY';
export const TX_WRAP_ETHER = TX__GROUP__WRAP_UNWRAP + 'WRAP';
export const TX_WRAP_TOKEN_WRAPPER = TX__GROUP__WRAP_UNWRAP + 'WRAP';
export const TX_UNWRAP_ETHER = TX__GROUP__WRAP_UNWRAP + 'UNWRAP_ETHER';
export const TX_UNWRAP_TOKEN_WRAPPER = TX__GROUP__WRAP_UNWRAP + 'UNWRAP_TOKEN_WRAPPER';


export const TX_STATUS_AWAITING_USER_ACCEPTANCE = 'TX/STATUS_AWAITING_USER_ACCEPTANCE'; // PENDING
export const TX_STATUS_AWAITING_CONFIRMATION = 'TX/STATUS_AWAITING_CONFIRMATION'; // PENDING
export const TX_STATUS_CONFIRMED = 'TX/STATUS_CONFIRMED';
export const TX_STATUS_CANCELLED_BY_USER = 'TX/STATUS_CANCELLED_BY_USER';
export const TX_STATUS_REJECTED = 'TX/STATUS_REJECTED';
export const TX_STATUS_ERROR = 'TX/STATUS_ERROR';

export const TX_ALLOWANCE_TRUST_TOGGLE = TX__GROUP__LIMITS + 'TX/ALLOWANCE_TRUST_TOGGLE';

export const TRANSACTION_IS_CONFIRMED = '0x1';
export const TRANSACTION_IS_REJECTED = '0x0';

export const getTransactionGroup = (transactionType) => {
  if (0 === transactionType.indexOf(TX__GROUP__OFFERS)) {
    return TX__GROUP__OFFERS;
  } else if (0 === transactionType.indexOf(TX__GROUP__TOKENS)) {
    return TX__GROUP__TOKENS;
  } else if (0 === transactionType.indexOf(TX__GROUP__LIMITS)) {
    return TX__GROUP__LIMITS;
  } else if (0 === transactionType.indexOf(TX__GROUP__TRANSFERS)) {
    return TX__GROUP__TRANSFERS;
  }
};

const init = createAction(
  'TX/INIT',
  () => null,
);

const transactionCancelledByUser = createAction(
  'TX/TRANSACTION_CANCELLED_BY_USER',
  ({ txType, txStatus, txStats }) => ({ txType, txStatus, txStats })
);

const addTransaction = createPromiseActions('TX/ADD_TRANSACTION');
const addTransactionEpic = (
  { txType, txHash, txMeta, txDispatchedTimestamp, txStartTimestamp },
  { latestBlockNumber = network.latestBlockNumber,
    sync = syncTransaction,
    transactionCheckInterval = transactions.transactionCheckInterval} = {}
  ) => (dispatch, getState) => {

  // console.log('addTransactionEpic');

  let previousBlockNumber = latestBlockNumber(getState());

  dispatch(
    addTransaction.pending({
      txHash, txType, txMeta,
      txStatus: TX_STATUS_AWAITING_CONFIRMATION,
      txStats: {
        txDispatchedTimestamp, txStartTimestamp,
        txStartBlockNumber: previousBlockNumber}
    }),
  );

  return new Promise((resolve, reject) => {
    const complete = setInterval(async () => {
      const currentBlockNumber = latestBlockNumber(getState());

      // console.log(currentBlockNumber, previousBlockNumber);

      if (currentBlockNumber !== previousBlockNumber) {

        previousBlockNumber = currentBlockNumber;

        let txReceipt;

        // console.log('before syncTransaction', txHash);

        try {
          txReceipt = (await dispatch(sync(txHash))).value;
        } catch (e) {
          //unlimited retries?
          console.warn('syncTransaction result', e);
          return;
        }

        // console.log('txReceipt', txReceipt);

        if (txReceipt == null) return;

        const confirmed = txReceipt.status === TRANSACTION_IS_CONFIRMED;

        // console.log('confirmed', confirmed);

        const payload = {
          txHash, txReceipt, txType,
          txConfirmedBlockNumber: txReceipt.blockNumber,
          txStatus: confirmed ? TX_STATUS_CONFIRMED : TX_STATUS_REJECTED,
          txGasCost: txReceipt.gasCost,
          txStats: {
            txEndBlockNumber: currentBlockNumber,
            txEndTimestamp: getTimestamp()
          }
        };

        // console.log('payload', payload);

        if(confirmed) {
          resolve(payload);
          dispatch(addTransaction.fulfilled(payload));
        } else {
          reject(payload);
          dispatch(addTransaction.rejected(payload));
        }
        clearInterval(complete)
      }
    }, transactionCheckInterval(getState()));
  });
};

const syncTransaction = createAction(
  'TX/SYNC_TRANSACTION',
  txHash => web3p.eth.getTransactionReceipt(txHash),
);

const transactionRejected = createAction(
  'TX/TRANSACTION_REJECTED',
  err => err,
);

const getCurrentGasPrice = createAction(
  'TX/GET_CURRENT_GAS_PRICE',
  async () => web3p.eth.getGasPrice(),
);

const getCurrentTxNonce = createAction(
  'TX/GET_CURRENT_TX_NONCE',
  async (accountAddress) => web3p.eth.getTransactionCount(accountAddress || web3.eth.defaultAccount)
);

const getCurrentTxNonceEpic = (address) => (dispatch, getState) => {
  const subjectAddress = address || accounts.defaultAccount(getState());
  dispatch(getCurrentTxNonce(subjectAddress));
};

const actions = {
  init,
  addTransactionEpic,
  transactionRejected,
  syncTransaction,
  getCurrentGasPrice,
  transactionCancelledByUser,
  getCurrentTxNonceEpic
};

const reducer = handleActions({
  [addTransaction.pending]:
    (
      state, { payload: { txHash, txType, txStats, txMeta } },
    ) => {
      const txPayload = fromJS({
        txHash,
        txReceipt: null,
        txType,
        txStatus : TX_STATUS_AWAITING_CONFIRMATION,
        txStats,
        txMeta
      });
      return state.updateIn(
        ['txList'], txList => txList.push(txPayload),
      );
    },
  [addTransaction.fulfilled]:
    (
      state, { payload: { txHash, txReceipt, txStatus, txStats } },
    ) => {
      const txListIndex = state.get('txList').findIndex(tx => tx.get('txHash') === txHash);
      return state.updateIn(
        ['txList', txListIndex],
        transaction => transaction
          .set('txReceipt', txReceipt)
          .set('txStatus', txStatus)
          .setIn(['txStats','txEndTimestamp'], txStats.txEndTimestamp)
          .setIn(['txStats','txEndBlockNumber'], txStats.txEndBlockNumber)
          .setIn(['txStats','txTotalTimeSec'], txStats.txEndTimestamp - transaction.getIn('txStartTimestamp'))
        ,
      );
    },
  [addTransaction.rejected]:
    (
      state, { payload: { txHash, txReceipt, txStatus, txStats } }
    ) => {

    const txListIndex = state.get('txList').findIndex(tx => tx.get('txHash') === txHash);
      return state.updateIn(
        ['txList', txListIndex],
        transaction => transaction
          .set('txReceipt', txReceipt)
          .set('txStatus', txStatus)
          .setIn(['txStats','txEndTimestamp'], txStats.txEndTimestamp)
          .setIn(['txStats','txEndBlockNumber'], txStats.txEndBlockNumber)
          .setIn(['txStats','txTotalTimeSec'], txStats.txEndTimestamp - transaction.getIn('txStartTimestamp'))
      );
  },
  [fulfilled(getCurrentTxNonce)]: (state, { payload }) => state.set('txNonce', parseInt(payload) + 1 ),
  [fulfilled(getCurrentGasPrice)]: (state, { payload }) => state.set('currentGasPriceInWei', payload.toString()),
}, initialState);

export default {
  actions,
  reducer,
};
