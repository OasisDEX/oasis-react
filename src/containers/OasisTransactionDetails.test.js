/* global shallow describe it expect */
/* eslint-disable import/first */
import React from 'react';
import { fromJS } from 'immutable';

import {
  OasisTransactionDetailsWrapper,
  mapStateToProps,
  mapDispatchToProps
} from './OasisTransactionDetails';
import { shallow } from 'enzyme';
import { TX_OFFER_TAKE } from '../store/reducers/transactions';
import { TOKEN_DAI, TOKEN_WRAPPED_ETH } from '../constants';

describe('(Container) OasisTransactionDetails', () => {
  const state = fromJS(global.storeMock);
  const initialProps = mapStateToProps(state, { transactionSubectType: TX_OFFER_TAKE });
  const initialActions = mapDispatchToProps(x => x);
  const props = {
    ...initialActions,
    ...initialProps,
    transactionSubectType: TX_OFFER_TAKE,
    getTransactionGasCostEstimate: jest.fn,
    buyToken: TOKEN_DAI,
    sellToken: TOKEN_WRAPPED_ETH,
    transactionSubjectAddress: '0x0000000000000000000000000000000000000000'
  };

  it('will receive right props', () => {
    expect(initialProps).toMatchSnapshot();
  });


  it('will receive right actions', () => {
    expect(initialActions).toMatchSnapshot();
  });

  it('should render', () => {
    const wrapper = shallow(
      <OasisTransactionDetailsWrapper {...props}/>
    );
    expect(wrapper).toMatchSnapshot();
  });

});
