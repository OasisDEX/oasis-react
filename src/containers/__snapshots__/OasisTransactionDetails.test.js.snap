// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Container) OasisTransactionDetails should render 1`] = `
<div
  style={
    Object {
      "backgroundColor": "lightgray",
      "margin": "20px 0",
      "padding": 10,
    }
  }
>
  <div>
    Invalid transaction
  </div>
</div>
`;

exports[`(Container) OasisTransactionDetails will receive right actions 1`] = `
Object {
  "actions": Object {},
}
`;

exports[`(Container) OasisTransactionDetails will receive right props 1`] = `
Object {
  "baseToken": undefined,
  "gasEstimateError": null,
  "isGasEstimatePending": null,
  "latestBlockNumber": 6716998,
  "transactionGasCostEstimate": null,
}
`;
